projects/router_kit/man/man8/rkd.8
.TH rkd 8 "December 2007" "NetFPGA2" "NetFPGA2 Manual"

.SH NAME
rkd \- Mirror Linux ipv4 forwarding state to NetFPGA reference router

.SH SYNOPSIS
.B rkd
[OPTIONS]

.SH DESCRIPTION
\fBrkd\fR mirrors Linux software forwarding state associated with
NetFPGA2 interfaces by reading \fB/proc/net/route\fR and
\fB/proc/net/arp\fR and writing to NetFPGA2's register interface.

.SH OPTIONS
.TP
.BR \-h ", " \-\^\-help
Prints a brief help message to the console.

.TP
.BR \-f ", " \-\^\-daemon
Run rkd in the background.

.SH COMMANDS
.TP
.BI \-p ", " \-\^\-interval " MS"
Specify the polling interval in milliseconds.

.SH BUGS
Currently doesn't take advantage of netlink for updates to routing table, but rather relies on
more expensive polling.


CHANGES
HEAD -- Version X.X.X beta

  Improvements:
   * Register system:
     - project name included in the #define at the top of the C header file,
       allowing the inclusion of multiple header files
     - device ID info (version, name, desc) now specified in project.xml
     - provide more information in headers at the top of generated files
     - output bitmasks in C/Perl
   * device_id module updated to include proj name, desc, and directory
   * nf2_info utility updated and installed by default (reads the device ID
     info from the active bitfile)
   * Moved contributed code out of lib/verilog/core
   * Added regression tests for the reference switch
   * Packaging scripts allow Verilog modules to be excluded
   * selftest updated to use new register system
   * Types added for CPCI registers with bitmasks
   * Added functions to C common library:
     - read device ID register
     - verify that the correct bitfile is downloaded
   * Self test software verifies that the selftest bitfile is downloaded before
     running the test
   * Kernel module:
     - added basic ethtool support (Thanks to Kumar Sanghvi)
     - added mii-tool support (Thanks to Kumar Sanghvi)
     - support newer Linux kernels (net_device API deprecated)
       (Thanks to Paul Rodman & Maciej Żenczykowski @ Google)
   * DRAM related:
     - DRAM block read/write test, DRAM queue test (1 queue), DRAM router 
       (8 queue) now released as full source code 
     - Several timing improvement to make the DRAM related modules faster 

  Bug fixes:
   * numerous fixes to register system:
     - incorrectly output of instance block addresses
     - replace periods with underscores in identifier names
   * cpci_download:
     - should work correctly on Fedora systems
       (Thanks to Paul Rodman)

  Other changes:
   * CPCI project renamed: CPCI_2.1 -> cpci
   * CPCI PERL library updated: CPCI_21Lib.pm -> CPCI_Lib.pm
   * Preparation for 10G system:
     - Perl NF2 library renamed to NF
     - nf2 replaced with nf in utility names (eg. nf_download)
     - NetFPGA directory renamed to netfpga (instead of NF2)

Version 2.0.1 beta
(29 Sep, 2009)

  Improvements:
   * Modified packaging script to support core and contibuted verilog libraries
   * Added regression tests to Reference Switch
   * Released Reference Switch

Version 2.0 beta
(30 July, 2009)

  Improvements:
   * Generate CPCI Registers when building a Project
   * JAVA GUI uses new Register System
   * Reference NIC regression tests rewritten
   * Support shared contants/types in XML

Version 2.0 alpha
(15 May, 2009)

 Improvements:
  * Register XML backend
  * Added nf2_register_gen.pl to generate registers
  * All scripts updated to create registers
  * All reference designs updated to use project.xml file
  * All library modules updated to use module xml files
  * Updated nf2_download (allows for link status)
  * DRAM Output FIFO
  * Reference NIC regression tests rewritten
  * Generic Table Registers (lib/verilog)

 Bug fixes
  * Rate limiter lockup condition fixed

Version 1.2.5-2 beta
(4 Feb, 2009)

 Bug Fixes
  * Fixed the regress.txt file located in NF2/projects

Version 1.2.5-1 beta
(22 Jan, 2009)

 Bug fixes
  * Replaced selftest bitfile to ensure it meets timing

Version 1.2.5 beta
(20 Jan, 2009)

 Bug fixes
  * MAC IO queues
    - Disabled MAC truncation of packets for ethertypes < 46 (not all modules are short-packet safe)
  * BRAM Output queues
    - Fixed bug with register counter updates

 Improvements:
  * Driver
   - DMA transfer errors include more debugging information


Version 1.2.4 beta
(11 Aug, 2008)

 Minor new features and improvements:
  * CPCI
   - Corrected bug preventing programming of Virtex via JTAG immediately after power-on

 Improvements:
  * Updated location information in UCF files for CPCI reprogramming signals
    (these are unused in all but the CPCI reprogrammer) -- eliminates warning
    about location constraints for signal pins


Version 1.2.3 beta
(21 Jul, 2008)

 Improvements:
  * Updated the reference NIC and reference router regression tests
  * Fixed the ucf files for projects (contributed from Jeff Shafer)


Version 1.2 beta revision 2
(10 Jul, 2008)

 Minor improvment:
  * CPCI
   - Increased the DMA default timeout value


Version 1.2 beta
(08 Jul, 2008)

 Major new features:
  * Packaging
   - RPM packages released to improve installation process for CentOS users

  * Updated DMA engine
    - DMA transfers from host to NetFPGA use the dedicated DMA bus between Spartan and Virtex
    - Supports operation at both 125 MHz and 62.5 MHz (previous version only functioned at 125 MHz)


 Minor new features and improvements:
  * Kernel driver
    - Checks added to prevent ioctl access outside of NetFPGA address space
    - Checks added to prevent ioctl calls prior to complete card initialization (crashes seen on some CentOS 5 systems)
    - Register read/write functions exported to allow other kernel modules to use NetFPGA module
    - Added tx packet/byte counters (bug #4)
    - Kernel unloading/reloading fixed (bug #15 + unreported bug)

  * nf2_download
    - Verifies that the Virtex bitfile was compiled against the CPCI version currently loaded (bug #54)

  * cpci_reprogram
    - Added --bit option to specify the CPCI bitfile to download

  * Selftest
    - Updated to support new DMA code
    - Register addresses reassigned

  * Utilities
    - Added nf2_info tool (lib/C/tools/nf2_info) to display info in the device_id module
    - Added sram_dump tool (lib/scripts/sram_dump) to dump the contents of SRAM

  * Perl libraries
    - Libraries restructed slightly to allow use in other projects (external interfaces should remain unchanged)

  * Compatibility
    - Minor changes to improve compatibility with non-supported Linux distributions

  * Projects
    - Project specific C/Perl register defines will be built when a local Verilog register defines exists



Version 1.1 beta
(29 Feb, 2008)


projects/router_kit/sw/rkd.cc
//-----------------------------------------------------------------------------
// File: rkd.cc
// Date: Sun Apr 29 16:59:49 PDT 2007
// Author: Martin Casado
//
// Description:
//
// Router kit user-level program for monitoring arouting and ARP table and
// syncronizing with NetFPGAv2.1 hardware.
//
//-----------------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <string>
#include <cstdio>
#include <cstdlib>

#include "iflist.hh"
#include "nf21_mon.hh"
#include "linux_proc_net.hh"

extern "C" {
#include <getopt.h>
#include <unistd.h>
#include <fcntl.h>
#include <signal.h>
}


using namespace rk;
using namespace std;

static int DEFAULT_INTERVAL_MS = 50; // Be default poll proc every 50ms

// --
// Vanilla daemonization code shamelessly adapted from
// (http://www.enderunix.org/docs/eng/daemon.php)
// --

void signal_handler(int sig)
{
    switch(sig) {
    case SIGHUP:
        break;
    case SIGTERM:
        ::exit(0);
        break;
    }
}

void daemonize()
{
    int i;
    if(::getppid()==1) return;
    i=::fork();
    if (i<0) ::exit(1);
    if (i>0) ::exit(0);
    ::setsid();
    for (i=::getdtablesize();i>=0;--i) ::close(i);
    i=::open("/dev/null",O_RDWR); ::dup(i); ::dup(i);
    ::umask(027);
    ::chdir("/tmp");
    ::signal(SIGCHLD,SIG_IGN);
    ::signal(SIGTSTP,SIG_IGN);
    ::signal(SIGTTOU,SIG_IGN);
    ::signal(SIGTTIN,SIG_IGN);
    ::signal(SIGHUP,signal_handler);
    ::signal(SIGTERM,signal_handler);
}

void usage(const string& argv0)
{
    cout << argv0 << " the NetFPGA2 linux shadowing daemon " << endl;
    cout << "usage: " << argv0 << " [-h|--help][-d|--daemon][-i|--interval=<poll interval>]" << endl;
    cout << "\t[-h|--help] : this message " << endl;
    cout << "\t[-d|--daemon] : run "<<argv0<<" a a daemon process " << endl;
    cout << "\t[-p|--interval]= : set the polling interval " << endl;
    cout << "\t[-i|--interface]= : set the NetFPGA base interface " << endl;
    ::exit(0);
}

std::string long_options_to_short_options(struct option options[])
{
    std::string short_options;
    for (; options->name; options++) {
        struct option* o = options;
        if (o->flag == NULL && o->val > 0 && o->val <= UCHAR_MAX) {
            short_options.push_back(o->val);
            if (o->has_arg == required_argument) {
                short_options.push_back(':');
            } else if (o->has_arg == optional_argument) {
                short_options.append("::");
            }
        }
    }
    return short_options;
}


int main(int argc,char **argv)
{
    bool daemon = false;
    int  interval  = DEFAULT_INTERVAL_MS;
    char interface[32];
    bzero(interface, 32);


    for (;;) {
        static struct option long_options[] = {
            {"help",  no_argument, 0, 'h'},
            {"daemon",  no_argument, 0, 'd'},

            {"interval",      required_argument, 0, 'p'},
            {"interface",      required_argument, 0, 'i'},
            {0, 0, 0, 0},
        };
        static std::string short_options(long_options_to_short_options(
                    long_options));
        int option_index;

        int c;

        c = getopt_long(argc, argv, short_options.c_str(),
                long_options, &option_index);
        if (c == -1)
            break;

        switch (c) {
            case 'h':
                usage(argv[0]);
                break;

            case 'd':
                daemon = true;
                break;

            case 'p':
                interval = atoi(optarg);
                break;

            case 'i':
                strncpy(interface, optarg, 32);
                break;

            case '?':
                exit(EXIT_FAILURE);

            default:
                abort();
        }

    }

    rtable rtcur;
    rtable rtcheck;

    iflist ifcur;
    iflist ifcheck;

    arptable arpcur;
    arptable arpcheck;

    nf21_mon eh_mon(interface);


    if(daemon){
        daemonize();
    }

    while(1)
    {
        linux_proc_net_load_rtable(rtcheck);

        if(rtcheck != rtcur){
            eh_mon.rtable_update(rtcheck);
            rtcur = rtcheck;
        }

        linux_proc_net_load_arptable(arpcheck);

        if(arpcheck != arpcur)
        {
            arpcur = arpcheck;
            eh_mon.arptable_update(arpcheck);
        }

        fill_iflist(ifcheck);

        if(ifcheck != ifcur){
            eh_mon.interface_update(ifcheck);
            ifcur = ifcheck;
        }

        ::usleep(interval);
    }

    return 0;
}
